File: skills/community-ecology-ordination/scripts/community_analysis.R
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

# Usage: Rscript community_analysis.R <species_site_matrix.csv> <metadata.csv> <output_dir> [method]
# NMDS ordination, diversity metrics, PERMANOVA
# Usage: Rscript community_analysis.R <species_matrix_csv> <metadata_csv> <output_dir>
# Requires: vegan, ggplot2, dplyr

# ── Inline logger ─────────────────────────────────────────────────────────────
SKILL_NAME <- "community-ecology-ordination"
.log_ts  <- function() format(Sys.time(), "[%Y-%m-%d %H:%M:%S]")
log_info <- function(...) message(.log_ts(), " [INFO]  ", sprintf(...))
log_warn <- function(...) message(.log_ts(), " [WARN]  ", sprintf(...))
log_error<- function(...) message(.log_ts(), " [ERROR] ", sprintf(...))
log_step <- function(n, d) log_info("-- STEP %d: %s", n, d)
log_decision <- function(v, val, why) log_info("DECISION | %s = %s | %s", v, val, why)
dir.create("logs", recursive=TRUE, showWarnings=FALSE)

suppressPackageStartupMessages({
  library(vegan)
  library(ggplot2)
  library(dplyr)
})

args       <- commandArgs(trailingOnly = TRUE)
sp_file    <- ifelse(length(args) >= 1, args[1], "data/species_matrix.csv")
meta_file  <- ifelse(length(args) >= 2, args[2], "data/site_metadata.csv")
output_dir <- ifelse(length(args) >= 3, args[3], "outputs/community")

log_step(1, "Validate inputs")
if (!file.exists(sp_file)) {
  log_error(
    "Failed in validate inputs: species matrix file not found: %s\nProbable cause: incorrect path or file not generated\nCheck: the species_matrix_csv argument and working directory\nPrevious skill: data-cleaning",
    sp_file
  )
  stop("Species matrix file not found.")
}
if (!file.exists(meta_file)) {
  log_error(
    "Failed in validate inputs: metadata file not found: %s\nProbable cause: incorrect path or file not generated\nCheck: the metadata_csv argument and working directory\nPrevious skill: data-cleaning",
    meta_file
  )
  stop("Metadata file not found.")
}

dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
set.seed(42)
log_decision("random_seed", 42, "ensures reproducibility of NMDS and permutation tests")

log_step(2, "Load species matrix and metadata")
tryCatch({
  sp   <- read.csv(sp_file, row.names = 1)
  meta <- read.csv(meta_file, row.names = 1)
}, error = function(e) {
  log_error(
    "Failed in load data: %s\nProbable cause: malformed CSV or missing rownames column\nCheck: file structure (first column must be site ID)\nPrevious skill: data-cleaning",
    conditionMessage(e)
  )
  stop(e)
})

log_info("Sites: %d | Species: %d", nrow(sp), ncol(sp))

if (any(sp < 0, na.rm = TRUE)) {
  log_warn("Species matrix contains negative values. Abundances must be >= 0. Check your input data.")
}
if (anyNA(sp)) {
  log_warn("Species matrix contains %d NA values. These will be treated as zero by vegan.", sum(is.na(sp)))
}

log_step(3, "Compute alpha diversity metrics")
tryCatch({
  div <- data.frame(
    site    = rownames(sp),
    richness = specnumber(sp),
    shannon  = diversity(sp, index = "shannon"),
    simpson  = diversity(sp, index = "simpson")
  )
  write.csv(div, file.path(output_dir, "diversity_metrics.csv"), row.names = FALSE)
  log_info("Alpha diversity computed. Mean richness: %.1f | Mean Shannon: %.2f",
           mean(div$richness), mean(div$shannon))
}, error = function(e) {
  log_error(
    "Failed in alpha diversity: %s\nProbable cause: empty or non-numeric species matrix\nCheck: structure of the species CSV\nPrevious skill: data-cleaning",
    conditionMessage(e)
  )
  stop(e)
})

log_step(4, "Run NMDS ordination (Bray-Curtis, k=2)")
log_decision("distance_metric", "bray", "Bray-Curtis is standard for community composition data")
log_decision("nmds_k", 2, "2 dimensions for interpretable 2D ordination plot")
tryCatch({
  nmds <- metaMDS(sp, distance = "bray", k = 2, trymax = 50, trace = 0)
  log_info("NMDS stress: %.4f", nmds$stress)
  if (nmds$stress > 0.2) {
    log_warn("NMDS stress = %.4f exceeds 0.20. Ordination may be unreliable; consider k=3 or data transformation.", nmds$stress)
  }

  scores_df <- as.data.frame(scores(nmds, display = "sites")) |>
    mutate(site = rownames(sp)) |>
    left_join(meta |> mutate(site = rownames(meta)), by = "site")

  p_ord <- ggplot(scores_df, aes(x = NMDS1, y = NMDS2)) +
    geom_point(size = 3) +
    annotate("text", x = Inf, y = -Inf, label = paste("Stress =", round(nmds$stress, 3)),
             hjust = 1.1, vjust = -0.5, size = 3.5) +
    theme_bw() + labs(title = "NMDS Ordination (Bray-Curtis)")
  ggsave(file.path(output_dir, "ordination_plot.png"), p_ord, width = 7, height = 6, dpi = 150)
  log_info("Ordination plot saved.")
}, error = function(e) {
  log_error(
    "Failed in NMDS: %s\nProbable cause: matrix with insufficient sites/species or all zeros\nCheck: number of sites (>= 3) and that the matrix is not all zeros\nPrevious skill: data-cleaning",
    conditionMessage(e)
  )
  stop(e)
})

log_step(5, "PERMANOVA and PERMDISP (if 'group' column present)")
if ("group" %in% names(meta)) {
  log_decision("permanova_permutations", 999, "standard number for robust p-value estimation")
  tryCatch({
    dist_mat  <- vegdist(sp, method = "bray")
    perm      <- adonis2(dist_mat ~ meta$group, permutations = 999)
    disp      <- betadisper(dist_mat, meta$group)
    disp_test <- permutest(disp, permutations = 999)
    log_info("PERMANOVA:\n%s", paste(capture.output(perm), collapse = "\n"))
    log_info("PERMDISP:\n%s", paste(capture.output(disp_test), collapse = "\n"))
    capture.output(perm, disp_test) |>
      writeLines(file.path(output_dir, "permanova_results.txt"))
    log_info("PERMANOVA results saved.")
  }, error = function(e) {
    log_error(
      "Failed in PERMANOVA/PERMDISP: %s\nProbable cause: group with only one level or insufficient sites per group\nCheck: 'group' column in metadata and balance\nPrevious skill: data-cleaning",
      conditionMessage(e)
    )
    stop(e)
  })
} else {
  log_warn("Column 'group' not found in metadata. PERMANOVA and PERMDISP skipped.")
}

log_info("Done. Outputs in: %s", output_dir)

File: skills/community-ecology-ordination/scripts/community_analysis.py
#!/usr/bin/env python3
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

"""
community_analysis.py
Beta diversity, ordination (PCoA), and group comparison (PERMANOVA via skbio).
Usage: python community_analysis.py <species_matrix_csv> <metadata_csv> <output_dir>
Requires: pandas, numpy, scipy, skbio, matplotlib
"""
import logging
import sys
from datetime import datetime
from pathlib import Path

SKILL_NAME = "community-ecology-ordination"
_LOG_DIR   = Path("logs")
_LOG_DIR.mkdir(parents=True, exist_ok=True)
_log_file  = _LOG_DIR / f"skill_{SKILL_NAME}_{datetime.now().strftime('%Y%m%d_%H%M%S')}.log"
logging.basicConfig(
    level=logging.INFO,
    format="[%(asctime)s] [%(levelname)s] [" + SKILL_NAME + "] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
    handlers=[
        logging.StreamHandler(sys.stdout),
        logging.FileHandler(_log_file, encoding="utf-8"),
    ],
)
logger = logging.getLogger(SKILL_NAME)

def log_step(n: int, desc: str) -> None:
    logger.info("-- STEP %d: %s", n, desc)

def log_decision(var: str, val, why: str) -> None:
    logger.info("DECISION | %s = %s | %s", var, val, why)

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.spatial.distance import braycurtis
from scipy.cluster.hierarchy import dendrogram, linkage, copshenetic
from scipy.spatial.distance import squareform

try:
    from skbio.diversity import beta_diversity
    from skbio.stats.ordination import pcoa
    from skbio.stats.distance import permanova, DistanceMatrix
    HAS_SKBIO = True
except ImportError:
    HAS_SKBIO = False
    logger.warning("scikit-bio not installed. PCoA and PERMANOVA will be skipped. pip install scikit-bio")


def bray_curtis_matrix(sp: pd.DataFrame) -> np.ndarray:
    n = len(sp)
    dm = np.zeros((n, n))
    vals = sp.values.astype(float)
    for i in range(n):
        for j in range(i+1, n):
            d = braycurtis(vals[i], vals[j])
            dm[i, j] = dm[j, i] = d
    return dm

def alpha_diversity(sp: pd.DataFrame) -> pd.DataFrame:
    richness = (sp > 0).sum(axis=1)
    def shannon(row):
        p = row[row > 0] / row.sum()
        return -np.sum(p * np.log(p))
    def simpson(row):
        p = row[row > 0] / row.sum()
        return 1 - np.sum(p**2)
    return pd.DataFrame({
        "site":     sp.index,
        "richness": richness.values,
        "shannon":  sp.apply(shannon, axis=1).values,
        "simpson":  sp.apply(simpson, axis=1).values,
    })

def main():
    sp_file    = sys.argv[1] if len(sys.argv) > 1 else "data/species_matrix.csv"
    meta_file  = sys.argv[2] if len(sys.argv) > 2 else "data/site_metadata.csv"
    output_dir = Path(sys.argv[3]) if len(sys.argv) > 3 else Path("outputs/community")

    log_step(1, "Validate inputs")
    if not Path(sp_file).exists():
        logger.error(
            "Species matrix file not found: %s\n"
            "Probable cause: incorrect path or file not yet generated\n"
            "Check: o argumento species_matrix_csv e o working directory\n"
            "Previous skill: data-cleaning",
            sp_file
        )
        sys.exit(1)
    if not Path(meta_file).exists():
        logger.error(
            "Metadata file not found: %s\n"
            "Probable cause: incorrect path or file not yet generated\n"
            "Check: o argumento metadata_csv e o working directory\n"
            "Previous skill: data-cleaning",
            meta_file
        )
        sys.exit(1)

    output_dir.mkdir(parents=True, exist_ok=True)

    log_step(2, "Load species matrix and metadata")
    try:
        sp   = pd.read_csv(sp_file, index_col=0)
        meta = pd.read_csv(meta_file, index_col=0)
    except Exception as e:
        logger.error(
            "Unexpected error in load data: %s\n"
            "Probable cause: CSV malformado ou sem coluna de rownames\n"
            "Check: structra dos arquivos (primeira coluna deve ser site ID)\n"
            "Previous skill: data-cleaning",
            e
        )
        raise

    logger.info("Sites: %d | Species: %d", len(sp), len(sp.columns))

    if (sp < 0).any().any():
        logger.warning("Species matrix contains negative values. Abundances must be >= 0. Check your input data.")
    if sp.isna().any().any():
        logger.warning("Species matrix contains %d NA values. These will affect distance calculations.", sp.isna().sum().sum())

    log_step(3, "Compute alpha diversity metrics")
    try:
        div = alpha_diversity(sp)
        div.to_csv(output_dir / "diversity_metrics.csv", index=False)
        logger.info("Mean richness: %.1f | Shannon: %.2f", div['richness'].mean(), div['shannon'].mean())
    except Exception as e:
        logger.error(
            "Unexpected error in alpha diversity: %s\n"
            "Probable cause: matriz de especies vazia ou nao numerica\n"
            "Check: CSV structure de especies\n"
            "Previous skill: data-cleaning",
            e
        )
        raise

    log_step(4, "Compute Bray-Curtis distance matrix")
    log_decision("distance_metric", "bray-curtis", "standard for community composition; handles double-zeros correctly")
    try:
        dm = bray_curtis_matrix(sp)
        pd.DataFrame(dm, index=sp.index, columns=sp.index).to_csv(output_dir / "bray_curtis_matrix.csv")
        logger.info("Bray-Curtis matrix computed (%d x %d).", len(sp), len(sp))
    except Exception as e:
        logger.error(
            "Unexpected error in Bray-Curtis matrix: %s\n"
            "Probable cause: dados nao numericos na matriz de especies\n"
            "Check: tipos de dados no CSV de especies\n"
            "Previous skill: data-cleaning",
            e
        )
        raise

    log_step(5, "PCoA ordination and PERMANOVA")
    if HAS_SKBIO:
        log_decision("permanova_permutations", 999, "standard number for robust p-value estimation")
        try:
            dist_mat = DistanceMatrix(dm, ids=list(sp.index))
            pc = pcoa(dist_mat)
            scores = pc.samples[["PC1", "PC2"]].copy()
            scores["site"] = scores.index
            if "group" in meta.columns:
                scores["group"] = meta["group"].reindex(scores.index).values
                groups_for_perm = meta["group"].reindex(sp.index).values
                perm_result = permanova(dist_mat, groups_for_perm, permutations=999)
                logger.info(
                    "PERMANOVA: F = %.3f | p = %.4f",
                    perm_result['test statistic'], perm_result['p-value']
                )
                perm_df = pd.DataFrame({"statistic": [perm_result["test statistic"]],
                                        "p_value":   [perm_result["p-value"]]})
                perm_df.to_csv(output_dir / "permanova_results.csv", index=False)
                # Plot coloured by group
                fig, ax = plt.subplots(figsize=(7, 6))
                for grp in scores["group"].unique():
                    sub = scores[scores["group"] == grp]
                    ax.scatter(sub["PC1"], sub["PC2"], label=grp, s=50, alpha=0.8)
                ax.set_xlabel(f"PC1 ({pc.proportion_explained[0]*100:.1f}%)")
                ax.set_ylabel(f"PC2 ({pc.proportion_explained[1]*100:.1f}%)")
                ax.set_title("PCoA (Bray-Curtis)")
                ax.legend(); plt.tight_layout()
                plt.savefig(output_dir / "pcoa_plot.png", dpi=150)
                plt.close()
                logger.info("PCoA plot saved.")
            else:
                logger.warning("Column 'group' not found in metadata. PERMANOVA skipped.")
        except Exception as e:
            logger.error(
                "Unexpected error in PCoA/PERMANOVA: %s\n"
                "Probable cause: grupo com apenas um nivel ou sites insuficientes\n"
                "Check: coluna 'group' nos metadados e balanceamento\n"
                "Previous skill: data-cleaning",
                e
            )
            raise
    else:
        logger.warning("scikit-bio unavailable. PCoA and PERMANOVA steps skipped.")

    log_step(6, "Hierarchical clustering")
    try:
        Z = linkage(squareform(dm), method="ward")
        c, _ = copshenetic(Z, squareform(dm))
        log_decision("linkage_method", "ward", "minimises total within-cluster variance; standard for ecology")
        logger.info("Cophenetic correlation (Ward): %.3f", c)
        if c < 0.7:
            logger.warning("Cophenetic correlation = %.3f < 0.70. Dendrogram may poorly represent distances.", c)
        fig, ax = plt.subplots(figsize=(max(8, len(sp)//2), 5))
        dendrogram(Z, labels=list(sp.index), ax=ax, leaf_rotation=90, leaf_font_size=8)
        ax.set_title(f"Hierarchical Clustering (Ward.D2) | Cophenetic r = {c:.3f}")
        plt.tight_layout()
        plt.savefig(output_dir / "cluster_dendrogram.png", dpi=150)
        plt.close()
        logger.info("Cluster dendrogram saved.")
    except Exception as e:
        logger.error(
            "Unexpected error in hierarchical clustering: %s\n"
            "Probable cause: distance matrix contains NaN or only one site\n"
            "Check: integridade da matriz Bray-Curtis\n"
            "Previous skill: community-ecology-ordination (distance matrix)",
            e
        )
        raise

    logger.info("Outputs written to: %s", output_dir)

if __name__ == "__main__":
    main()
